### PR TITLE
Fix initrand to avoid random number sequences overlapping

### LINKED ISSUE #17898: `randomPathName` called twice in a row can return the same string on windows
this is the 2nd bug with std/tempfiles i've encountered while working on https://github.com/nim-lang/Nim/pull/17892 (i already sent a fix for the 1st one https://github.com/nim-lang/Nim/issues/17888 which affected it). This shows that dog-fooding is useful and we should do more of it (ie, promote using stdlib within nim repo).

### Example
```nim
# see https://github.com/nim-lang/Nim/pull/17892 for `genTempPath`, which
# is based on `randomPathName`
proc diffStrings*(a, b: string): tuple[output: string, same: bool] =
  template tmpFileImpl(prefix, str): auto =
    let path = genTempPath(prefix, "")
    writeFile(path, str)
    path
  let patha = tmpFileImpl("diffStrings_a_", a)
  let pathb = tmpFileImpl("diffStrings_b_", b)
  defer:
    removeFile(patha)
    removeFile(pathb)
  result = diffFiles(patha, pathb)
```

### Current Output
on windows, in CI, `echo diffStrings(lhs, rhs)` produced the following:
```
2021-04-29T20:00:01.1496001Z --- "a/C:\\Users\\VSSADM~1\\AppData\\Local\\Temp\\diffStrings_a_C5xLA9E2"
2021-04-29T20:00:01.1507701Z +++ "b/C:\\Users\\VSSADM~1\\AppData\\Local\\Temp\\diffStrings_b_C5xLA9E2"
```
(see also https://gist.github.com/timotheecour/515dde92cb29ce359a3f4b9546c247fd)

it looks like `randomPathName` produced the same string `C5xLA9E2`, and this is unlikely to be by accident.

### Expected Output
`randomPathName` called twice in a row should return different strings with very high probability.

### Additional Information
1.5.1 16405083485967a395b8d677833bc26040881b21

I'm not on windows, but windows users should be able to minimize this issue with a reproducible example based on this, eg by calling `randomPathName` in a loop and checking for nearest neighbor duplicates

/cc @xflywind can you reproduce this on your end?

### PR DESCRIPTION
**Pull request by demotomohiro**
This is the alternative of https://github.com/nim-lang/Nim/pull/18149.
Fix https://github.com/nim-lang/Nim/issues/17898.
It allows 2^64 `Rand` generate 2^64 random numbers without random number sequences overlapping.

If states of PRNGs were initialized using time, multiple states of PRNG (pseudorandom number generator) can have same value when they are initialized at almost same time.
If states of PRNG were initialized with random value, sequences generated by them can be overlapped.
https://blogs.sas.com/content/iml/2018/05/07/overlap-random-streams.html
The probability of overlap of random subsequences of pseudorandom number generator is explained in this paper:
https://vigna.di.unimi.it/ftp/papers/overlap.pdf

According to the paper, the probability of overlap is always at most n^2 * L / P where n is a number of PRNG, L is a length of sequences and P is a period of the PRNG.

This PR implementates `initRand` in the same way as runnableExamples of `skipRandomNumbers` in `random` module.
It try to avoid a PRNG state identicals to a state of other PRNGs in other processes by initializing the `baseState` with `std/sysrand`.
When it doesn't worked, `baseSeed` is initialized with hash value of current process id and monotime.
But different processes in different machines can have same process id.

`initRand` avoids random number sequences overlapping by copying `baseState` to new `Rand` state and call `skipRandomNumbers` with `baseState`.

I didn't touched js code because how `skipRandomNumbers` works in js is unknown.
Even if it works and can skip 2^64 times `next()` calls, that means it is same to call next() once because maximum period of `Rand` in js is 2^64 - 1, because `sizeof(Rand) == 8` in js.
I think it should be replaced with 32-bit Generators like xoshiro128++ or xoshiro128** if random module doesn't need to generate same sequences in all platform.
https://prng.di.unimi.it
Or implement 64bit int addition and bit operation with 32 bit int for correct xoroshiro128+.

I visualized how Pseudo-random number generators created with jump function(`skipRandomNumbers`) and with random seed use PRNG state space.

prng
Source:
https://gist.github.com/demotomohiro/9516fe686da376bbce8a4470e52752ab

Each blue box represents a PRNG state.
Red rounded box represents PRNG that generates 4 random numbers by updating state 4 times.
Upper block represents how PRNGs created with jump function uses PRNG states.
Lower block represents how PRNGs created with random seed uses PRNG states.
In lower block, several PRNGs uses same states.
That means a part of random number sequence from one PRNG matchs a part of random number sequence from other PRNG.

This document explains how this PR avoid random number sequences overlapping:
https://www.mathcha.io/editor/7JPgyt0WSLdH4L7VywudZkBrQF4qJOe8Fp3JBB

### CODE REVIEW

**Comment by timotheecour** [comment]
> If states of PRNGs were initialized using time, multiple states of PRNG (pseudorandom number generator) can have same value when they are initialized at almost same time.

that's not true anymore, I fixed this in https://github.com/nim-lang/Nim/pull/17468 which already calls `skipRandomNumbers(result)` in the right place; it's unclear whether your PR adds redundant calls to `skipRandomNumbers` given that `skipRandomNumbers` is already being called by `initRand(seed)`

File: lib/std/tempfiles.nim [hunk 1]
```diff
@@ -107,11 +107,8 @@ var nimTempPathState {.threadvar.}: NimTempPathState
 template randomPathName(length: Natural): string =
   var res = newString(length)
   if not nimTempPathState.isInit:
-    var time = getMonoTime().ticks
```
Comment (ringabout) [inline lib/std/tempfiles.nim:110]: or just `xor getProcessorId()`.

And remove `import std/monotimes`

**Comment by demotomohiro** [comment]
> > If states of PRNGs were initialized using time, multiple states of PRNG (pseudorandom number generator) can have same value when they are initialized at almost same time.
> 
> that's not true anymore, I fixed this in #17468 which already calls `skipRandomNumbers(result)` in the right place; it's unclear whether your PR adds redundant calls to `skipRandomNumbers` given that `skipRandomNumbers` is already being called by `initRand(seed)`

`initRand*(seed: int64)` is a function because it always returns same value from same input.
If `initRand*(seed: int64): Rand` was called with `times.getTime()` twice and first and second `times.getTime()` call returned the same value as first `times.getTime()`,  two `Rand` returned from `initRand*()` would have same status.

In this PR, one process have only one `baseState` that is initialized on start up with random number from `sysrand.urandom` or monotime + process Id.
After `baseState` was initialized, it is updated only with `skipRandomNumbers` in `initRand()`.
When `initRand()` is called, `baseState` is copied to new `Rand` and `baseState` is updated with `skipRandomNumbers`.
`initRand()` doesn't take random value from time or sysrand and doesn't call `initRand*(seed: int64)` except js backend.
So When `initRand()` is called multiple times, states of `Rand` returned from `initRand()` doesn't overlap to each other unless it calls `next*(r: var Rand): uint64` more than 2^64 times or `initRand` is called more than 2^64 times.

This document explains how this PR avoid random number sequences overlapping:
https://www.mathcha.io/editor/7JPgyt0WSLdH4L7VywudZkBrQF4qJOe8Fp3JBB

File: lib/std/tempfiles.nim (hunk 1 above)
Comment (demotomohiro) [reply lib/std/tempfiles.nim:110]: Thank you!
I have removed `import std/monotimes`.
I think `initRand() xor getProcessorId()` is same to initializing `Rand` with random value and doesn't prevent random number sequences overlapping.
I think `xor getProcessorId().hash()` to each output of `Rand` would be better to avoid overlapping to outputs in other process as `f(x) = x xor a` is bijection.
But that would require adding a new feature to random module.

File: lib/pure/random.nim [hunk 2]
```diff
@@ -617,15 +622,60 @@ proc shuffle*[T](x: var openArray[T]) =
 
   shuffle(state, x)
 
-when not defined(nimscript) and not defined(standalone):
-  import times
+when not defined(standalone):
+  when defined(nimscript):
```
Comment (Varriount) [inline lib/pure/random.nim]: This would be slightly better (in my mind) as
```
when defined(standalone):
  discard
elif defined(nimscript):
  ...
```

File: lib/pure/random.nim [hunk 3]
```diff
@@ -617,15 +622,60 @@ proc shuffle*[T](x: var openArray[T]) =
...
+        ret.a1 = t.hash().Ui
+        if not ret.isValid:
+          ret.a0 = pid.Ui
+          ret.a1 = t.Ui
+
+      if not ret.isValid:
+        ret = DefaultRandSeed
+      ret
+
+    when compileOption("threads"):
```
Comment (Varriount) [inline lib/pure/random.nim]: Wouldn't it be better to have the default random state be a per-thread global? That way, locking wouldn't be needed.

File: lib/pure/random.nim [hunk 4]
```diff
@@ -617,15 +622,60 @@ proc shuffle*[T](x: var openArray[T]) =
...
+    import std/hashes
+
+    var baseState = block:
+      var ret = Rand(
+        a0: CompileTime.hash.Ui,
+        a1: CompileDate.hash.Ui)
+      if not ret.isValid:
+        ret = DefaultRandSeed
+      ret
+  elif defined(js):
```
Comment (Varriount) [inline lib/pure/random.nim]: As long as it's equivalent, I believe this is slightly simpler:
```suggestion
    var baseState = Rand(
      a0: CompileTime.hash.Ui,
      a1: CompileDate.hash.Ui
    )
    if not ret.isValid:
      baseState = DefaultRandSeed

  elif defined(js):
```

File: lib/pure/random.nim [hunk 5]
```diff
@@ -617,15 +622,60 @@ proc shuffle*[T](x: var openArray[T]) =
...
+    import std/times
+  else:
+    import std/[hashes, os, sysrand, monotimes]
+
+    var baseState: Rand = block:
```
Comment (timotheecour) [inline lib/pure/random.nim]: this is problematic, `import rand` shouldn't trigger code that executes at the top-level; it should be lazily initialized on 1st use

File: lib/pure/random.nim [hunk 6]
```diff
@@ -617,15 +622,60 @@ proc shuffle*[T](x: var openArray[T]) =
...
+        # When 2 processes executed at same time on different machines,
+        # `ret` can still have same value.
+        let
+          pid = getCurrentProcessId()
+          t = getMonoTime().ticks
+        ret.a0 = pid.hash().Ui
+        ret.a1 = t.hash().Ui
+        if not ret.isValid:
+          ret.a0 = pid.Ui
+          ret.a1 = t.Ui
```
Comment (Varriount) [inline lib/pure/random.nim]: When would the data supplied by `sysrand.urandom` not correctly initialize the global random state?

**Review by Varriount — CHANGES_REQUESTED** [review]

File: lib/pure/random.nim (hunk 5 above)
Comment (demotomohiro) [reply lib/pure/random.nim]: Thank you! I fixed it.
`baseState` is initialized in initRand() proc.

File: lib/pure/random.nim (hunk 2 above)
Comment (demotomohiro) [reply lib/pure/random.nim]: This code cannot be made simpler in such a way as `initRand()` and `randomize()` are defined under the `when not defined(standalone):`.

File: lib/pure/random.nim (hunk 3 above)
Comment (demotomohiro) [reply lib/pure/random.nim]: `baseState` cannot be a per-thread global because if it was per-thread, it cannot avoid random number sequences overlapping between different PRNGs in different thread.

Quote from https://prng.di.unimi.it (web site created by creator of xoroshiro):
> We provide ready-made jump functions for a number of calls equal to the square root of the period, to make it easy generating non-overlapping sequences for parallel computations, and equal to the cube of the fourth root of the period, to make it possible to generate independent sequences on different parallel processors.

`skipRandomNumbers*(s: var Rand)` in random module is a jump function.

File: lib/pure/random.nim (hunk 4 above)
Comment (demotomohiro) [reply lib/pure/random.nim]: Thank you!
I changed code to that style.

File: lib/pure/random.nim (hunk 6 above)
Comment (demotomohiro) [reply lib/pure/random.nim]: xoroshiro128+ doesn't work when all bits of state are zero.
If `sysrand.urandom` is not guaranteed to never return zeros when sizeof(Rand) bytes are taken from it, we need to check returned value.
Maybe there is a bug in OS and `sysrand.urandom` returns zero.
Even if `sysrand.urandom` is working correctly, you might get a zero unfortunately (probability of 128bits random number returns 0 is 1/2^128 ).

According to the comment of `sysrand.urandom` proc,
> If the call succeeds, returns `true`.

In other words, it can fail. So we need alternative ways to get random number.
Even if probability of `sysrand.urandom` returns 128bit zeros is tiny, it is easy to take care of such case if we already have alternative ways to get random number.

File: lib/pure/random.nim [hunk 7]
```diff
@@ -635,20 +653,60 @@ when not defined(nimscript) and not defined(standalone):
...
+              copyMem(result.addr, urand[0].addr, sizeof(Rand))
+              if result.isValid:
+                break
+
+          if not result.isValid:
+            # When 2 processes executed at same time on different machines,
+            # `result` can still have same value.
+            let
+              pid = getCurrentProcessId()
+              t = getMonoTime().ticks
```
Comment (Araq) [inline lib/pure/random.nim]: If `urandom` failed 8 times in a row, just `quit` instead, don't continue after a bug.

File: lib/pure/random.nim [hunk 8]
```diff
@@ -635,20 +653,60 @@ when not defined(nimscript) and not defined(standalone):
...
+            result.a0 = pid.hash().Ui
+            result.a1 = t.hash().Ui
+            if not result.isValid:
+              result.a0 = pid.Ui
+              result.a1 = t.Ui
```
Comment (Araq) [inline lib/pure/random.nim]: This code was probably never tested and is a liability for security.

File: lib/pure/random.nim [hunk 9]
```diff
@@ -635,20 +653,60 @@ when not defined(nimscript) and not defined(standalone):
       let time = int64(times.epochTime() * 1000) and 0x7fff_ffff
       result = initRand(time)
     else:
-      let now = times.getTime()
-      result = initRand(convert(Seconds, Nanoseconds, now.toUnix) + now.nanosecond)
+      proc getRandomState(): Rand =
```
Comment (timotheecour) [inline lib/pure/random.nim:660]: move this nested proc to top-level scope, simpler/cleaner

File: lib/pure/random.nim (hunk 7 above)
Comment (timotheecour) [reply lib/pure/random.nim]: instead of quit, how about raise a new `RandError`, which user can then handle and use their preferred fallback; shouldn't be a concern in practice I'm assuming?

File: lib/pure/random.nim [hunk 10]
```diff
@@ -635,20 +653,60 @@ when not defined(nimscript) and not defined(standalone):
...
+        when defined(nimscript):
```
Comment (timotheecour) [inline lib/pure/random.nim:661]: how about adding a vmops for this, that way you don't need to special case nimscript with a suboptimal seeding, and it then also works for VM

can be done in followup PR, alternatively

File: tests/stdlib/trandom.nims [hunk 11]
```diff
@@ -0,0 +1,12 @@
+import std/[random, sets]
```
Comment (timotheecour) [inline tests/stdlib/trandom.nims]: optional:
instead how about
```
# in tests/test_nimscript.nims
import stdlib/trandom
```
and use conditional `when defined(nimscript)` as needed to make tests pass

which increases test coverage without duplication

File: lib/pure/random.nim (hunk 6 above)
Comment (Varriount) [reply lib/pure/random.nim]: What about `AND`ing the state with `1` after seeding via urandom? That would ensure that the state is never zero.
Alternately, an explicit check and retry would work.

File: lib/pure/random.nim (hunk 3 above)
Comment (Varriount) [reply lib/pure/random.nim]: I don't follow - if per-thread state is initialized with random data via urandom, why would `skipRandomNumbers` overlap?

File: lib/pure/random.nim (hunk 8 above)
Comment (demotomohiro) [reply lib/pure/random.nim]: So `baseState` should be initialized only with `sysrand.urandom` and if it failed, just quit or raise error?

File: lib/pure/random.nim (hunk 9 above)
Comment (demotomohiro) [reply lib/pure/random.nim:660]: Why moving nested proc to top-level scope makes simpler/cleaner?
It doesn't reduce a total number of lines of random module.
`getRandomState` proc is used only in `initRand()`.

File: tests/stdlib/trandom.nims (hunk 11 above)
Comment (demotomohiro) [reply tests/stdlib/trandom.nims]: I have removed `tests/stdlib/trandom.nims` and added `import stdlib/trandom` to `tests/test_nimscript.nims`.

File: lib/pure/random.nim (hunk 3 above)
Comment (demotomohiro) [reply lib/pure/random.nim]: Please read this:
https://www.mathcha.io/editor/7JPgyt0WSLdH4L7VywudZkBrQF4qJOe8Fp3JBB

Initial state of all `Rand` must be T2^64n (n is natural number including 0) to avoid overlapping.
If `baseState` were per-thread variable and they were initialized with `urandom`, initial state of all `Rand` cannot become T2^64n.

File: lib/pure/random.nim (hunk 6 above)
Comment (demotomohiro) [reply lib/pure/random.nim]: `isValid(r: Rand): bool` func checks whether state is zero or not.
```nim
func isValid(r: Rand): bool {.inline.} =
  not (r.a0 == 0 and r.a1 == 0)
```
This PR retrys `urandom` if `isValid` was false.

File: lib/pure/random.nim (hunk 7 above)
Comment (Araq) [reply lib/pure/random.nim]: And who knows about this lovely new exception. And how are you supposed to handle it? It's simply "we're screwed completely". The worst thing you can do after a bug is to continue after it because then you're in an unknown, untested state. That's a *very different* thing from a "restart after bug" as the restart starts in the program state that has seen most testing by design, the initial state.

File: lib/pure/random.nim (hunk 8 above)
Comment (Araq) [reply lib/pure/random.nim]: Yes, just quit.

File: lib/pure/random.nim (hunk 10 above)
Comment (Araq) [reply lib/pure/random.nim:661]: Yeah, followup PR, it's controversial.

File: lib/pure/random.nim (hunk 8 above)
Comment (demotomohiro) [reply lib/pure/random.nim]: Done!

File: lib/pure/random.nim (hunk 7 above)
Comment (demotomohiro) [reply lib/pure/random.nim]: It `quit` when `urandom` failed 8 times in a row now.

File: tests/stdlib/trandom.nim [hunk 12]
```diff
@@ -248,3 +249,26 @@ block: # bug #17670
...
+      randSet.incl rands[i]
+
+    doAssert randSet.len == size
+
+    # Checks random number sequences overlapping.
+    const numRepeat = 100
+    for i in 0..<size:
+      for j in 0..<numRepeat:
+        discard rands[i].next
+        doAssert rands[i] notin randSet
```
Comment (Araq) [inline tests/stdlib/trandom.nim:274]: Should be addressed in a follow up PR but the notion that a random sequence of numbers doesn't "overlap" with some other sequence seems to weaken the "randomness" and not to improve it.

File: lib/pure/random.nim (hunk 3 above)
Comment (planetis-m) [reply lib/pure/random.nim]: @demotomohiro ...so now every call to next needs to hold the lock! I imagine this would be very slow. Is there a better solution than ignoring the data races?

Comment (Varriount) [reply lib/pure/random.nim]: So if I understand correctly, you're stating that the problem is that `Rand` only has a single set of states (T0 to T2^64n), and that the initial seed only determines which of the 2^64n is the initial state of a particular `Rand` instance? (I'm not well-versed in how random number generation works).

The description at https://prng.di.unimi.it/ implies that the xoroshiro algorithm *should* have characteristics that make it feasible to produce non-overlapping sequences in parallel situations. I'm also not sure that "doesn't produce overlapping sequences" is a guarantee that the random library should provide.

File: tests/stdlib/trandom.nim (hunk 12 above)
Comment (timotheecour) [reply tests/stdlib/trandom.nim:274]: i don't think think this is correct; the sequences of numbers generated can have overlap in parts, eg:
1 (9 8 3) 2 5
2 (9 8 3) 7 1

but the underlying states won't overlap (so long we remain within a period); if they did, the sequences generated would be identical at any point after which the state overlaps

File: lib/pure/random.nim (hunk 3 above)
Comment (timotheecour) [reply lib/pure/random.nim]: @demotomohiro I'm also concerned about the point regarding performance cause by required locking raised in https://github.com/nim-lang/Nim/pull/18744#discussion_r701046207 ; could you please run some benchmarks to measure this?

note that the approach i proposed in https://github.com/nim-lang/Nim/pull/18149 did not have such problem

Comment (Varriount) [reply lib/pure/random.nim]: @timotheecour The approach in https://github.com/nim-lang/Nim/pull/18149 would suffer from overlap issues though, wouldn't it? Which is what I believe @demotomohiro is aiming to avoid.

Comment (timotheecour) [reply lib/pure/random.nim]: > The approach in #18149 would suffer from overlap issues though, wouldn't it?

I was also ensuring `skipRandomNumbers` gets called but indeed, the way it was being called wouldn't guarantee lack of overlap ; i'll need to think more to double check;

that said, the potential performance issue should be evaluated/investigated

Comment (Varriount) [reply lib/pure/random.nim]: Actually, upon re-reviewing the code:
  - The lock is only for initialization of new random states (from`basestate`), including the global state. I had previously misread this, and thought all access to the global random state (`state`) was being locked.
  - The current global random state isn't threadsafe at all (unless the implementation is a lock-free algorithm of some sort), and isn't a per-thread global. An obvious way to solve this to make `state` a thread-local global, and check/initialize it when it is accessed. This would impose a performance cost, though I don't know how small it would be - it would probably depend on whether the backend compiler is smart enough to hoist the random state somewhere on the stack.

Comment (demotomohiro) [reply lib/pure/random.nim]: @planetis-m
As @Varriount said, only `initRand()` proc acquires lock.
If you create `Rand` in each threads and never share them between threads, you don't need to lock when you call `next(r: var Rand)`.
If you create only one `Rand` and want to share it by multiple threads, you would need to acquire lock when you call `next(r: var Rand)`,
but it is not a problem this PR cause.

@Varriount 
In this document, T0 is a initial state of `baseState` and it is the value came from `sysrand.urandom`. 
https://www.mathcha.io/editor/7JPgyt0WSLdH4L7VywudZkBrQF4qJOe8Fp3JBB
Then, first `Rand` returned from `initRand()` has state T0.
second one has T2^64, third one has T2^64*2, 4th one has T2^64*3, n-th one has T2^64*(n-1) and so on.
Then, first `Rand` returned from `initRand()` can use states from T0 to T2^64-1 without overlapping the state with second one.
Second one can use from T2^64 to T2^64*2-1, third one can use from T2^64*2 to T2^64*3-1, and so on.

@timotheecour
I don't think performance of `initRand()` is so important, because it would not be called as frequently as `next(r: var Rand)` as `initRand()` is called only when you create a new `Rand`.
When many `Rand` are created but they generate few random numbers, performance of `initRand()` can be important.
But in that case, I think hash functions tested in paper would be better:
http://jcgt.org/published/0009/03/02/

> An obvious way to solve this to make `state` a thread-local global, and check/initialize it when it is accessed.

I think making `state` a thread-local variable and Nim compiler automatically inserts a call to `randomize()` so that it is called when a new thread starts would be better.
But I don't know it is possible.

File: tests/stdlib/trandom.nim (hunk 12 above)
Comment (demotomohiro) [reply tests/stdlib/trandom.nim:274]: Could you tell me why is it weaken randomness?
It just works like evenly cut a random number sequence generated by calling `next(r: var Rand)` with single `Rand`, and assign each sub sequence to each `Rand`.
If you think randomness of random module is not good enough, I think we need to use different algorithm like xoroshiro128++, xoroshiro128** or PCG.

File: lib/pure/random.nim (hunk 3 above)
Comment (planetis-m) [reply lib/pure/random.nim]: I get it now, good one +1. The key thing to remember is `state` and `baseState` are different variables. `state` is used only for the "simple" not thread safe api, while `baseState` is used in the new initRand overload.
